File: packages/forma-python/src/forma/provider.py
import json
import os
import urllib.request
from pathlib import Path
from typing import Callable, Protocol

from .types import FormaValue


class PermissionTools(Protocol):
    def require(self, permission: str) -> None:
        ...

    def read_text(self, path: str) -> str:
        ...

    def search_text(self, query: str) -> list[str]:
        ...

    def run_test(self, command: str) -> dict[str, object]:
        ...

    def write_text(self, path: str, content: str) -> dict[str, object]:
        ...


class ModelProvider(Protocol):
    def run_agent(
        self,
        instruction: str,
        values: dict[str, FormaValue],
        permissions: list[str],
        tools: PermissionTools,
        output: dict[str, dict[str, object]] | None = None,
        schemas: dict[str, dict[str, dict[str, object]]] | None = None,
    ) -> dict[str, FormaValue]:
        ...


class StaticProvider:
    def __init__(self, output: dict[str, FormaValue]) -> None:
        self.output = output

    def run_agent(
        self,
        instruction: str,
        values: dict[str, FormaValue],
        permissions: list[str],
        tools: PermissionTools,
        output: dict[str, dict[str, object]] | None = None,
        schemas: dict[str, dict[str, dict[str, object]]] | None = None,
    ) -> dict[str, FormaValue]:
        return self.output


Transport = Callable[[str, dict[str, object], dict[str, str]], dict[str, object]]


class HttpJsonProvider:
    def __init__(
        self,
        endpoint: str,
        model: str,
        api_key: str | None = None,
        transport: Transport | None = None,
    ) -> None:
        self.endpoint = endpoint
        self.model = model
        self.api_key = api_key
        self.transport = transport or self._default_transport

    def run_agent(
        self,
        instruction: str,
        values: dict[str, FormaValue],
        permissions: list[str],
        tools: PermissionTools | None,
        output: dict[str, dict[str, object]] | None = None,
        schemas: dict[str, dict[str, dict[str, object]]] | None = None,
    ) -> dict[str, FormaValue]:
        headers = {"content-type": "application/json"}
        if self.api_key:
            headers["authorization"] = f"Bearer {self.api_key}"

        tool_results: list[dict[str, object]] = []
        for _ in range(8):
            body = {
                "model": self.model,
                "instruction": instruction,
                "input": values,
                "permissions": permissions,
            }
            if tool_results:
                body["toolResults"] = tool_results
            response = self.transport(self.endpoint, body, headers)
            output = response.get("output")
            if isinstance(output, dict):
                return output
            calls = _parse_tool_calls(response.get("toolCalls"))
            if not calls:
                raise ValueError("F5001: provider response requires object output")
            if tools is None:
                raise ValueError("F4002: provider requested a host tool that is not configured")
            tool_results = [_run_tool_call(call, tools) for call in calls]
        raise ValueError("F5002: provider exceeded tool call limit")

    @staticmethod
    def _default_transport(url: str, body: dict[str, object], headers: dict[str, str]) -> dict[str, object]:
        request = urllib.request.Request(
            url,
            data=json.dumps(body).encode("utf8"),
            headers=headers,
            method="POST",
        )
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read().decode("utf8"))


class OpenAIResponsesProvider:
    def __init__(
        self,
        api_key: str,
        model: str,
        endpoint: str = "https://api.openai.com/v1/responses",
        transport: Transport | None = None,
    ) -> None:
        self.api_key = api_key
        self.model = model
        self.endpoint = endpoint
        self.transport = transport or self._default_transport

    def run_agent(
        self,
        instruction: str,
        values: dict[str, FormaValue],
        permissions: list[str],
        tools: PermissionTools | None,
        output: dict[str, dict[str, object]] | None = None,
        schemas: dict[str, dict[str, dict[str, object]]] | None = None,
    ) -> dict[str, FormaValue]:
        response = self.transport(
            self.endpoint,
            {
                "model": self.model,
                "instructions": instruction,
                "input": json.dumps({"input": values, "permissions": permissions}),
                "text": {
                    "format": {
                        "type": "json_schema",
                        "name": "forma_output",
                        "strict": True,
                        "schema": _object_schema(output or {}, schemas or {}),
                    }
                },
            },
            {
                "content-type": "application/json",
                "authorization": f"Bearer {self.api_key}",
            },
        )
        parsed_output = json.loads(_extract_output_text(response))
        if not isinstance(parsed_output, dict):
            raise ValueError("F5001: provider response requires object output")
        return parsed_output

    @staticmethod
    def _default_transport(url: str, body: dict[str, object], headers: dict[str, str]) -> dict[str, object]:
        request = urllib.request.Request(
            url,
            data=json.dumps(body).encode("utf8"),
            headers=headers,
            method="POST",
        )
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read().decode("utf8"))


ProviderProfile = dict[str, str]


def provider_profile_from_file(path: str | Path) -> ProviderProfile:
    parsed = json.loads(Path(path).read_text(encoding="utf8"))
    return _validate_provider_profile(parsed)


def provider_from_profile(
    profile: ProviderProfile,
    *,
    env: dict[str, str] | None = None,
    transport: Transport | None = None,
) -> ModelProvider:
    env_values = env or os.environ
    api_key = profile.get("apiKey") or (env_values.get(profile["apiKeyEnv"]) if "apiKeyEnv" in profile else None)
    provider = profile["provider"]
    if provider == "http-json":
        endpoint = profile.get("endpoint")
        if endpoint is None:
            raise ValueError("provider profile endpoint is required for http-json")
        return HttpJsonProvider(
            endpoint=endpoint,
            model=profile["model"],
            api_key=api_key,
            transport=transport,
        )
    if api_key is None:
        raise ValueError("provider profile apiKey or apiKeyEnv is required for openai-responses")
    return OpenAIResponsesProvider(
        api_key=api_key,
        model=profile["model"],
        endpoint=profile.get("endpoint", "https://api.openai.com/v1/responses"),
        transport=transport,
    )


def _validate_provider_profile(value: object) -> ProviderProfile:
    if not isinstance(value, dict):
        raise ValueError("provider profile must be a JSON object")
    provider = value.get("provider")
    if provider not in {"http-json", "openai-responses"}:
        raise ValueError("provider profile provider must be http-json or openai-responses")
    if not isinstance(value.get("model"), str) or not value["model"]:
        raise ValueError("provider profile model is required")
    for field in ["endpoint", "apiKey", "apiKeyEnv"]:
        if field in value and not isinstance(value[field], str):
            raise ValueError(f"provider profile {field} must be a string")
    return value


def _object_schema(fields: dict[str, dict[str, object]], schemas: dict[str, dict[str, dict[str, object]]]) -> dict[str, object]:
    return {
        "type": "object",
        "additionalProperties": False,
        "properties": {name: _field_schema(field, schemas) for name, field in fields.items()},
        "required": [name for name, field in fields.items() if not field["optional"]],
    }


def _parse_tool_calls(value: object) -> list[dict[str, object]]:
    if value is None:
        return []
    if not isinstance(value, list):
        raise ValueError("F5001: provider toolCalls must be a list")
    calls = []
    for item in value:
        if not isinstance(item, dict):
            raise ValueError("F5001: provider toolCalls must be objects")
        if not isinstance(item.get("id"), str) or not isinstance(item.get("name"), str):
            raise ValueError("F5001: provider toolCalls require id and name")
        if item["name"] not in {"readText", "searchText", "runTest", "writeText"}:
            raise ValueError(f"F5001: unsupported provider tool call '{item['name']}'")
        args = item.get("args", {})
        if not isinstance(args, dict):
            args = {}
        calls.append({"id": item["id"], "name": item["name"], "args": args})
    return calls


def _run_tool_call(call: dict[str, object], tools: PermissionTools) -> dict[str, object]:
    try:
        name = str(call["name"])
        args = call["args"]
        if not isinstance(args, dict):
            args = {}
        if name == "readText":
            result = tools.read_text(_string_arg(name, args, "path"))
        elif name == "searchText":
            result = tools.search_text(_string_arg(name, args, "query"))
        elif name == "runTest":
            result = tools.run_test(_string_arg(name, args, "command"))
        else:
            result = tools.write_text(_string_arg(name, args, "path"), _string_arg(name, args, "content"))
        return {"id": str(call["id"]), "ok": True, "result": result}
    except Exception as error:
        return {"id": str(call["id"]), "ok": False, "error": str(error)}


def _string_arg(tool_name: str, args: dict[object, object], name: str) -> str:
    value = args.get(name)
    if not isinstance(value, str):
        raise ValueError(f"F5001: provider tool call '{tool_name}' requires string arg '{name}'")
    return value


def _field_schema(field: dict[str, object], schemas: dict[str, dict[str, dict[str, object]]]) -> dict[str, object]:
    if field.get("array"):
        return {"type": "array", "items": _field_schema({**field, "array": False}, schemas)}
    field_type = str(field["type"])
    if field_type == "Text":
        return {"type": "string"}
    if field_type == "Number":
        return {"type": "number"}
    if field_type == "Boolean":
        return {"type": "boolean"}
    schema = schemas.get(field_type)
    return _object_schema(schema, schemas) if schema is not None else {}


def _extract_output_text(response: dict[str, object]) -> str:
    output_text = response.get("output_text")
    if isinstance(output_text, str):
        return output_text
    output = response.get("output")
    if not isinstance(output, list):
        raise ValueError("F5001: provider response requires output text")
    for item in output:
        if not isinstance(item, dict):
            continue
        content = item.get("content")
        if not isinstance(content, list):
            continue
        for part in content:
            if isinstance(part, dict) and isinstance(part.get("text"), str):
                return str(part["text"])
    raise ValueError("F5001: provider response requires output text")

File: packages/forma-typescript/src/provider.ts
import { readFileSync } from "node:fs";
import type { FormaField, FormaValue } from "./types.js";

export interface PermissionTools {
  require(permission: string): void;
  readText(path: string): Promise<string>;
  searchText(query: string): Promise<string[]>;
  runTest(command: string): Promise<{ ok: boolean; output: string }>;
  writeText(path: string, content: string): Promise<{ ok: boolean; output: string }>;
}

export interface ModelProvider {
  runAgent(input: {
    instruction: string;
    values: Record<string, FormaValue>;
    permissions: string[];
    output: Record<string, FormaField>;
    schemas: Record<string, Record<string, FormaField>>;
    tools: PermissionTools;
  }): Promise<Record<string, FormaValue>>;
}

export class StaticProvider implements ModelProvider {
  constructor(private readonly output: Record<string, FormaValue>) {}

  async runAgent(): Promise<Record<string, FormaValue>> {
    return this.output;
  }
}

type FetchLike = (
  url: string,
  init: {
    method: "POST";
    headers: Record<string, string>;
    body: string;
    signal?: AbortSignal;
  },
) => Promise<{ ok: boolean; status: number; text(): Promise<string> }>;

export interface HttpJsonProviderOptions {
  endpoint: string;
  model: string;
  apiKey?: string;
  temperature?: number;
  timeoutMs?: number;
  fetch?: FetchLike;
}

export class HttpJsonProvider implements ModelProvider {
  private readonly fetchImpl: FetchLike;

  constructor(private readonly options: HttpJsonProviderOptions) {
    this.fetchImpl = options.fetch ?? fetch;
  }

  async runAgent(input: {
    instruction: string;
    values: Record<string, FormaValue>;
    permissions: string[];
    output: Record<string, FormaField>;
    schemas: Record<string, Record<string, FormaField>>;
    tools: PermissionTools;
  }): Promise<Record<string, FormaValue>> {
    const headers: Record<string, string> = {
      "content-type": "application/json",
    };
    if (this.options.apiKey) {
      headers.authorization = `Bearer ${this.options.apiKey}`;
    }

    const toolResults: ToolResult[] = [];
    for (let attempt = 0; attempt < 8; attempt += 1) {
      const response = await this.fetchImpl(this.options.endpoint, {
        method: "POST",
        headers,
        ...timeoutSignal(this.options.timeoutMs),
        body: JSON.stringify({
          model: this.options.model,
          instruction: input.instruction,
          input: input.values,
          permissions: input.permissions,
          ...(this.options.temperature !== undefined ? { temperature: this.options.temperature } : {}),
          ...(toolResults.length > 0 ? { toolResults } : {}),
        }),
      });
      const text = await response.text();
      if (!response.ok) {
        throw new Error(`F5000: provider request failed with status ${response.status}`);
      }

      const parsed = JSON.parse(text) as { output?: unknown; toolCalls?: unknown };
      if (parsed.output && typeof parsed.output === "object" && !Array.isArray(parsed.output)) {
        return parsed.output as Record<string, FormaValue>;
      }
      const calls = parseToolCalls(parsed.toolCalls);
      if (calls.length === 0) {
        throw new Error("F5001: provider response requires object output");
      }
      toolResults.length = 0;
      for (const call of calls) {
        toolResults.push(await runToolCall(call, input.tools));
      }
    }
    throw new Error("F5002: provider exceeded tool call limit");
  }
}

interface ToolCall {
  id: string;
  name: "readText" | "searchText" | "runTest" | "writeText";
  args: Record<string, unknown>;
}

interface ToolResult {
  id: string;
  ok: boolean;
  result?: unknown;
  error?: string;
}

function parseToolCalls(value: unknown): ToolCall[] {
  if (!Array.isArray(value)) return [];
  return value.map((item): ToolCall => {
    if (!item || typeof item !== "object" || Array.isArray(item)) {
      throw new Error("F5001: provider toolCalls must be objects");
    }
    const call = item as { id?: unknown; name?: unknown; args?: unknown };
    if (typeof call.id !== "string" || typeof call.name !== "string") {
      throw new Error("F5001: provider toolCalls require id and name");
    }
    if (call.name !== "readText" && call.name !== "searchText" && call.name !== "runTest" && call.name !== "writeText") {
      throw new Error(`F5001: unsupported provider tool call '${call.name}'`);
    }
    return {
      id: call.id,
      name: call.name,
      args: call.args && typeof call.args === "object" && !Array.isArray(call.args) ? call.args as Record<string, unknown> : {},
    };
  });
}

async function runToolCall(call: ToolCall, tools: PermissionTools): Promise<ToolResult> {
  try {
    if (call.name === "readText") {
      return { id: call.id, ok: true, result: await tools.readText(stringArg(call, "path")) };
    }
    if (call.name === "searchText") {
      return { id: call.id, ok: true, result: await tools.searchText(stringArg(call, "query")) };
    }
    if (call.name === "runTest") {
      return { id: call.id, ok: true, result: await tools.runTest(stringArg(call, "command")) };
    }
    return {
      id: call.id,
      ok: true,
      result: await tools.writeText(stringArg(call, "path"), stringArg(call, "content")),
    };
  } catch (error) {
    return { id: call.id, ok: false, error: error instanceof Error ? error.message : String(error) };
  }
}

function stringArg(call: ToolCall, name: string): string {
  const value = call.args[name];
  if (typeof value !== "string") {
    throw new Error(`F5001: provider tool call '${call.name}' requires string arg '${name}'`);
  }
  return value;
}

export interface OpenAIResponsesProviderOptions {
  apiKey: string;
  model: string;
  endpoint?: string;
  temperature?: number;
  timeoutMs?: number;
  fetch?: FetchLike;
}

export class OpenAIResponsesProvider implements ModelProvider {
  private readonly fetchImpl: FetchLike;
  private readonly endpoint: string;

  constructor(private readonly options: OpenAIResponsesProviderOptions) {
    this.fetchImpl = options.fetch ?? fetch;
    this.endpoint = options.endpoint ?? "https://api.openai.com/v1/responses";
  }

  async runAgent(input: {
    instruction: string;
    values: Record<string, FormaValue>;
    permissions: string[];
    output: Record<string, FormaField>;
    schemas: Record<string, Record<string, FormaField>>;
    tools: PermissionTools;
  }): Promise<Record<string, FormaValue>> {
    const response = await this.fetchImpl(this.endpoint, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${this.options.apiKey}`,
      },
      ...timeoutSignal(this.options.timeoutMs),
      body: JSON.stringify({
        model: this.options.model,
        instructions: input.instruction,
        input: JSON.stringify({
          input: input.values,
          permissions: input.permissions,
        }),
        ...(this.options.temperature !== undefined ? { temperature: this.options.temperature } : {}),
        text: {
          format: {
            type: "json_schema",
            name: "forma_output",
            strict: true,
            schema: objectSchema(input.output, input.schemas),
          },
        },
      }),
    });
    const text = await response.text();
    if (!response.ok) {
      throw new Error(`F5000: provider request failed with status ${response.status}`);
    }
    const parsed = JSON.parse(text) as Record<string, unknown>;
    const outputText = extractOutputText(parsed);
    const output = JSON.parse(outputText) as unknown;
    if (!output || typeof output !== "object" || Array.isArray(output)) {
      throw new Error("F5001: provider response requires object output");
    }
    return output as Record<string, FormaValue>;
  }
}

export interface ProviderProfile {
  provider: "http-json" | "openai-responses";
  endpoint?: string;
  model: string;
  apiKey?: string;
  apiKeyEnv?: string;
  temperature?: number;
  timeoutMs?: number;
}

export interface ProviderProfileOptions {
  env?: Record<string, string | undefined>;
  fetch?: FetchLike;
}

export function providerProfileFromFile(path: string): ProviderProfile {
  const parsed = JSON.parse(readFileSync(path, "utf8")) as unknown;
  return validateProviderProfile(parsed);
}

export function providerFromProfile(profile: ProviderProfile, options: ProviderProfileOptions = {}): ModelProvider {
  const apiKey = profile.apiKey ?? (profile.apiKeyEnv ? (options.env ?? process.env)[profile.apiKeyEnv] : undefined);
  if (profile.provider === "http-json") {
    if (!profile.endpoint) {
      throw new Error("provider profile endpoint is required for http-json");
    }
    return new HttpJsonProvider({
      endpoint: profile.endpoint,
      model: profile.model,
      ...(apiKey ? { apiKey } : {}),
      ...(profile.temperature !== undefined ? { temperature: profile.temperature } : {}),
      ...(profile.timeoutMs !== undefined ? { timeoutMs: profile.timeoutMs } : {}),
      ...(options.fetch ? { fetch: options.fetch } : {}),
    });
  }
  if (!apiKey) {
    throw new Error("provider profile apiKey or apiKeyEnv is required for openai-responses");
  }
  return new OpenAIResponsesProvider({
    apiKey,
    model: profile.model,
    ...(profile.endpoint ? { endpoint: profile.endpoint } : {}),
    ...(profile.temperature !== undefined ? { temperature: profile.temperature } : {}),
    ...(profile.timeoutMs !== undefined ? { timeoutMs: profile.timeoutMs } : {}),
    ...(options.fetch ? { fetch: options.fetch } : {}),
  });
}

function validateProviderProfile(value: unknown): ProviderProfile {
  if (!value || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("provider profile must be a JSON object");
  }
  const profile = value as Partial<ProviderProfile>;
  if (profile.provider !== "http-json" && profile.provider !== "openai-responses") {
    throw new Error("provider profile provider must be http-json or openai-responses");
  }
  if (typeof profile.model !== "string" || profile.model.length === 0) {
    throw new Error("provider profile model is required");
  }
  if (profile.endpoint !== undefined && typeof profile.endpoint !== "string") {
    throw new Error("provider profile endpoint must be a string");
  }
  if (profile.apiKey !== undefined && typeof profile.apiKey !== "string") {
    throw new Error("provider profile apiKey must be a string");
  }
  if (profile.apiKeyEnv !== undefined && typeof profile.apiKeyEnv !== "string") {
    throw new Error("provider profile apiKeyEnv must be a string");
  }
  if (profile.temperature !== undefined && (typeof profile.temperature !== "number" || !Number.isFinite(profile.temperature))) {
    throw new Error("provider profile temperature must be a number");
  }
  if (profile.timeoutMs !== undefined && (typeof profile.timeoutMs !== "number" || !Number.isFinite(profile.timeoutMs) || profile.timeoutMs <= 0)) {
    throw new Error("provider profile timeoutMs must be a positive number");
  }
  return profile as ProviderProfile;
}

function timeoutSignal(timeoutMs: number | undefined): { signal?: AbortSignal } {
  if (timeoutMs === undefined) return {};
  return { signal: AbortSignal.timeout(timeoutMs) };
}

function objectSchema(fields: Record<string, FormaField>, schemas: Record<string, Record<string, FormaField>>): Record<string, unknown> {
  return {
    type: "object",
    additionalProperties: false,
    properties: Object.fromEntries(Object.entries(fields).map(([name, field]) => [name, fieldSchema(field, schemas)])),
    required: Object.entries(fields)
      .filter(([, field]) => !field.optional)
      .map(([name]) => name),
  };
}

function fieldSchema(field: FormaField, schemas: Record<string, Record<string, FormaField>>): Record<string, unknown> {
  if (field.array) {
    return {
      type: "array",
      items: fieldSchema({ ...field, array: false }, schemas),
    };
  }
  if (field.type === "Text") return { type: "string" };
  if (field.type === "Number") return { type: "number" };
  if (field.type === "Boolean") return { type: "boolean" };
  const schema = schemas[field.type];
  return schema ? objectSchema(schema, schemas) : {};
}

function extractOutputText(response: Record<string, unknown>): string {
  if (typeof response.output_text === "string") {
    return response.output_text;
  }
  const output = response.output;
  if (!Array.isArray(output)) {
    throw new Error("F5001: provider response requires output text");
  }
  for (const item of output) {
    if (!item || typeof item !== "object") continue;
    const content = (item as { content?: unknown }).content;
    if (!Array.isArray(content)) continue;
    for (const part of content) {
      if (part && typeof part === "object" && typeof (part as { text?: unknown }).text === "string") {
        return (part as { text: string }).text;
      }
    }
  }
  throw new Error("F5001: provider response requires output text");
}
